runtime/lua/vim/lsp/completion.lua

```diff
@@ -463,6 +463,188 @@ local function request(clients, bufnr, win, ctx, callback)
   end
 end
 
+---@param bufnr integer
+---@return string
+local function get_augroup(bufnr)
+  return string.format('nvim.lsp.completion_%d', bufnr)
+end
+
+---@param winid integer
+---@param bufnr integer
+---@param ft? string
+local function update_popup_window(winid, bufnr, ft)
+  if winid and api.nvim_win_is_valid(winid) and bufnr and api.nvim_buf_is_valid(bufnr) then
+    vim.wo[winid].conceallevel = 2
+    if ft then
+      local ok = pcall(vim.treesitter.get_parser, bufnr, ft)
+      if ok then
+        vim.treesitter.start(bufnr, vim.treesitter.language.get_lang(ft))
+      else
+        vim.wo[winid].eventignorewin = 'OptionSet'
+        vim.bo[bufnr].filetype = ft
+        vim.bo[bufnr].syntax = 'on'
+      end
+    end
+    local all = api.nvim_win_text_height(winid, {}).all
+    api.nvim_win_set_height(winid, all)
+  end
+end
+
+---@return function
+local function cmpitem_resolve()
```

Comment:
this is 100 lines of complex code. how is anyone supposed to know what this does without studying it?